javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -23,21 +23,34 @@
  * @since 2.0.0
  */
 public final class Vector<T> implements Kind1<Vector<?>, T>, IndexedSeq<T>, Serializable {
-
     private static final long serialVersionUID = 1L;
 
     private static final Vector<?> EMPTY = new Vector<>(HashArrayMappedTrie.empty());
 
-    private final HashArrayMappedTrie<Integer, T> trie;
-    private final int indexShift;
+    private final int offset, length;
```

Comment:
Until now the `indexShift` was only used as an offset for indices, i.e. if you deleted the first element, the first index became `1`, so we had to subtract `1` for every index. Now the underlying `HashArrayMappedTrie` remains the same (i.e. structural sharing), and the internal indexing can be negative also (see prepend)